src/ir/auth_logic/ast_construction.cc

```diff
@@ -0,0 +1,173 @@
+//-------------------------------------------------------------------------------
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+//-------------------------------------------------------------------------------
+
+#include <string>
+#include <vector>
+#include <utility>
+#include <stdexcept>
+#include <iostream>
+#include <memory>
+#include <variant>
+#include <optional>
+
+
+#include "antlr4-runtime.h"
+#include "absl/status/statusor.h"
+#include "src/common/utils/map_iter.h"
+#include "src/ir/auth_logic/ast.h"
+#include "src/ir/auth_logic/ast_construction.h"
+#include "src/policy/auth_logic/auth_logic_cc_generator_grammar.inc/AuthLogicLexer.h" 
+#include "src/policy/auth_logic/auth_logic_cc_generator_grammar.inc/AuthLogicParser.h"
+#include "src/policy/auth_logic/auth_logic_cc_generator_grammar.inc/AuthLogicVisitor.h"
+#include "src/ir/datalog/program.h"
+
+namespace raksha::ir::auth_logic {
+
+Principal construct_principal (AuthLogicParser::PrincipalContext* context) {
+    std::string name_ = context->ID()->getText();
+    return Principal(name_);
+}
+
+datalog::Predicate construct_predicate (AuthLogicParser::PredicateContext* context ) {
+    std::string name_ = context->ID()->getText();
+    std::vector<std::string> args_;
+    datalog::Sign sign_;
+    size_t index = 1;
+    for (auto& statement : context->pred_arg()) {
+        args_.push_back(std::move(statement->ID()->getText()));
+        index += 1;
+    }
+
+    if (context->NEG() == nullptr) {
+        sign_ = datalog::Sign::kPositive;
+    }
+    else {
+        sign_ = datalog::Sign::kNegated;
+    }
```

Comment:
It looks like you could declare and initialize the value if you user ternary operator:

```C++
auto sign = context->NEG() == nullptr ? datalog::Sign::kPositive: datalog::Sign::kNegative;
```